File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/two/Long_ScreenTwo_Stoch_D_TwoBelow_X.java
package com.kishlaly.ta.analyze.tasks.blocks.two;

import com.kishlaly.ta.analyze.tasks.ThreeDisplays;
import com.kishlaly.ta.analyze.tasks.blocks.TaskBlock;
import com.kishlaly.ta.model.BlockResultJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.model.indicators.IndicatorJava;
import com.kishlaly.ta.model.indicators.StochJava;
import com.kishlaly.ta.utils.LogJava;

import java.util.List;

import static com.kishlaly.ta.model.BlockResultCodeJava.OK;
import static com.kishlaly.ta.model.BlockResultCodeJava.STOCH_D_NOT_EXTRA_OVERSOLD_SCREEN_2;

/**
 * the last two %D stochastics below ThreeDisplays.Config.STOCH_CUSTOM
 */
public class Long_ScreenTwo_Stoch_D_TwoBelow_X implements TaskBlock {
    @Override
    public BlockResultJava check(SymbolDataJava screen) {
        List<StochJava> screen_2_Stochastic = (List<StochJava>) screen.indicators.get(IndicatorJava.STOCH);
        StochJava stoch2 = screen_2_Stochastic.get(screen_2_Stochastic.size() - 2);
        StochJava stoch1 = screen_2_Stochastic.get(screen_2_Stochastic.size() - 1);

        if (ThreeDisplays.Config.STOCH_CUSTOM < 0) {
            throw new RuntimeException("ThreeDisplays.Config.STOCH_CUSTOM not set");
        }

        boolean bothBelowExtraLow = stoch2.getSlowD() < ThreeDisplays.Config.STOCH_CUSTOM && stoch1.getSlowD() < ThreeDisplays.Config.STOCH_CUSTOM;

        if (!bothBelowExtraLow) {
            LogJava.recordCode(STOCH_D_NOT_EXTRA_OVERSOLD_SCREEN_2, screen);
            LogJava.addDebugLine("Both last %D stochastics are at least " + ThreeDisplays.Config.STOCH_CUSTOM + " on the second screen");
            return new BlockResultJava(screen.getLastQuote(), STOCH_D_NOT_EXTRA_OVERSOLD_SCREEN_2);
        }
        return new BlockResultJava(screen.getLastQuote(), OK);
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/two/stoch/Long_ScreenTwo_Stoch_D_TwoBelow_X.kt
package com.kishlaly.ta.analyze.tasks.blocks.two.stoch

import com.kishlaly.ta.analyze.tasks.ThreeDisplays
import com.kishlaly.ta.analyze.tasks.ThreeDisplays.Config.STOCH_CUSTOM
import com.kishlaly.ta.analyze.tasks.blocks.two.ScreenTwoBlock
import com.kishlaly.ta.model.BlockResult
import com.kishlaly.ta.model.BlockResultCode
import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.model.indicators.Stochastic
import com.kishlaly.ta.utils.Log
import com.kishlaly.ta.utils.SymbolDataUtils

/**
 * the last two %D stochastics below ThreeDisplays.Config.STOCH_CUSTOM
 */
class Long_ScreenTwo_Stoch_D_TwoBelow_X : ScreenTwoBlock {
    override fun check(screen: SymbolData): BlockResult {
        if (STOCH_CUSTOM < 0) {
            throw RuntimeException("ThreeDisplays.Config.STOCH_CUSTOM not set")
        }
        val stoch = SymbolDataUtils(screen, Stochastic::class.java)
        val bothBelowExtraLow = stoch.last(2).slowD < STOCH_CUSTOM
                && stoch.last(1).slowD < STOCH_CUSTOM

        if (!bothBelowExtraLow) {
            Log.recordCode(BlockResultCode.STOCH_D_NOT_EXTRA_OVERSOLD_SCREEN_2, screen)
            Log.addDebugLine("Both last %D stochastics are at least " + ThreeDisplays.Config.STOCH_CUSTOM + " on the second screen")
            return BlockResult(screen.lastQuote, BlockResultCode.STOCH_D_NOT_EXTRA_OVERSOLD_SCREEN_2)
        }

        return BlockResult(screen.lastQuote, BlockResultCode.OK)
    }
}